### ISSUE TITLE
Documents API: Concurrency validation is not backwards compatible (xAPI Version header, config.yml)

### ISSUE THREAD

**Issue by RoboSparrow** [Brightcookie/lxHive#23]
Submitted requests who are requiring eTags in xAPI 1.03. should not require this:

- when sent as former xAPI versions.
- the LRS is configured for support lower than 1.0.3
- when sent in legacy (IE) mode (POST ?method=<method>)

Currently any affected request are validated as 1.0.3, regardless of `X-Experience-API-Version` header or supported version settings in the Config.yml

Endpoints:

```
/agents
/activities/state
/activities
```

Config.yml, set to 1.0.1 but still having issue:
```
name: myGPcommunity
mode: development
database:
    host_uri: 'mongodb://127.0.0.1'
    db_name: lxHive
xAPI:
    latest_version: 1.0.1
    statement_get_limit: 100
    default_statement_get_format: exact
    supported_versions: [1.0.1,1.0.0]
    supported_auth_scopes: [{ name: super, description: 'Administrate the LRS' }, { name: statements/read, description: 'Read statements' }, { name: statements/read/mine, description: 'Read own statements' }, { name: attachments, description: 'Include attachments' }, { name: state, description: 'Access the State API' }, { name: profile, description: 'Access the Profile API' }, { name: define, description: '(Re)define new Activities' }, { name: all/read, description: 'Full read-only access' }, { name: all, description: 'Full access' }]
    oauth: { reauthorize_time: 0, reauthenticate_time: 2592000, branding: { enabled: false, logo_path: null, header: null, footer: null, css_path: null } }
filesystem:
    exposed_url: /attachments
    in_use: local
    local: { root_dir: '%app.root%/storage/files' }
```
**Also: we need to honour the [legacy request mode](https://github.com/adlnet/xAPI-Spec/blob/1.0.3/xAPI.md#alt-request-syntax) in the same way:**

i.e `POST https://api.experience.at/activities/state?method=GET` (returns a 500 with irrelevant CORS error message)

ADL example of CORS PUT request: https://github.com/adlnet/xAPI-Spec/blob/1.0.3/xAPI.md#appendix-g-cross-domain-request-example

Since applications like storyline send in legacy mode and *don't supply a version param* we have in this case to revert to the minimal supported version as defined in config.yml

@kienv @MarcusFabriczy can give you more details

**Comment by RoboSparrow** [comment]
```
curl 'http://dev.experience.at/agents/profile?agent=%7B%22mbox%22%3A%22mailto%3Ajasmine%40brightcookie.com.au%22%2C%22name%22%3A%22Jasmine%20Brightcookie%22%7D&profileId=Jasmine-1-18%2F8%2F2015%2012%3A04%3A10%20pm' -X POST -H 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8' -H 'Accept-Encoding: gzip, deflate' -H 'Accept-Language: en-US,en;q=0.5' -H 'Authorization: Basic MWxRdjN3eHB6WkUwSG9uN2l6STdYeEVhT1VLRUpEOE43WTNZd3NlVDpiVGY3TWdZUXdlNkV3Rjd3WjZJTzY1ZXRsVENrNnFTRlFFTHp6Z3RH' -H 'Content-Length: 59' -H 'Content-Type: application/json; charset=UTF-8' -H 'Host: dev.experience.at' -H 'Origin: http://devel.bo' -H 'Referer: http://devel.bo/lxJasmine/Jasmine/SpecRunner.Agents.html' -H 'User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:40.0) Gecko/20100101 Firefox/40.0' -H 'X-Experience-API-Version: 1.0.1'
```

returns:

 `409 Conflict` , `{"error_message":"There was a conflict. Check the current state of the resource and set the \"If-Match
\" header with the current ETag to resolve the conflict."}`

1.0.3: PUT, POST or DELETE
1.0.2: PUT only
1.0.1: PUT only

**Comment by sraka1** [comment]
I agree with points 1. and 2., however not with 3.
You can send any headers you want with the extra legacy layer as well. I know however, that some activity providers (namely Storyline) don't do this and as such are violating the spec.

In this case...I think it'd be best if there'd be some sort of extra middleware layer for such cases.

**Comment by sraka1** [comment]
Defaulting to a minimal supported version is problematic. Why? Because when we have i.e. versions 1.0 and 1.1 available, what if a request without a version comes in and we default to 1.0, but that request is actually a 1.1 request?

Providing a version is an important part of the xAPI spec.
See https://github.com/adlnet/xAPI-Spec/blob/master/xAPI.md#client-requirements-3

**Comment by sraka1** [comment]
My fault, I had to re-re-read the spec file. The spec actually says "SHOULD*" (so a SHOULD, which will probably become a must in 1.1.0) for POST and DELETE in 1.0.3. Fixed in #28.

Regarding the version header, I believe that Storyline sends this one way or another - either as a header or as a POST parameter using the legacy layer as other also LRS'es reject requests without a version.

Closing this issue.
